### PR TITLE
Need proper validation in model path

### PR DESCRIPTION
**Pull request by dpanshug**
Closes: [RHOAIENG-3314](https://issues.redhat.com/browse/RHOAIENG-3314)

## Description
Added condition to throw error with path is invalid, also reduced the field debounce timeout to 500ms

<details><summary>Valid path</summary>

image
</details>

<details><summary>Path with special chars</summary>

image
</details>

<details><summary>Path with only slash</summary>

Screenshot from 2024-03-04 15-01-15
</details>

<details><summary>Path with multiple slashes in between</summary>

image
</details>

<details><summary>Path starting with multiple slashes are invalid</summary>

image
</details>

## How Has This Been Tested?
1. Open deploy model
2. Enter path with different special chars and check the error.

## Test Impact
Added cypress test to check path with different values
Added unit test for string util function

## Request review criteria:

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit tests & storybook for related changes)

If you have UI changes: 

- [x] Included any necessary screenshots or gifs if it was a UI change.
- [x] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

@vconzola

### CODE REVIEW

**Comment by dpanshug** [comment]
@vconzola can you review this pls.

**Comment by vconzola** [comment]
> Path starting with slashes are valid, but the slashes will be removed after submitting
This is a nit, but for the above case, if there's only one leading slash, it should be ignored and removed after submitting. If there are multiple leading slashes, that should be an "Invalid path format." Also, note, there should be a period after the word "format" in this error text.

File: frontend/src/pages/modelServing/screens/projects/InferenceServiceModal/DataConnectionFolderPathField.tsx [hunk 1]
```diff
@@ -22,20 +23,30 @@ const DataConnectionFolderPathField: React.FC<DataConnectionFolderPathFieldProps
...
+      setValidated('error');
+    } else {
+      setValidated('default');
+    }
+  }, [folderPath]);
+
+  const debouncedValidateFolderPath = useDebounceCallback(validateFolderPath, 500);
+
+  React.useEffect(() => {
+    debouncedValidateFolderPath();
```
Comment (lucferbux) [inline frontend/src/pages/modelServing/screens/projects/InferenceServiceModal/DataConnectionFolderPathField.tsx:42]: Mmmmm I have a problem with this, I'll try to find some time to suggest a solution, let me tell you what's happening.

We use the debouncer to make sure users can input a leading slash (which btw we don't error out but it should be an error) and not error out due to the path being only a single slash.

The main problem with this debouncer as it is right now is that it willl trigger the error no matter what, see error bellow.

Ideally, we would want to cancel the previous debounce if the error is dismissed. The current approach works great and it's not a big of a deal, but let's see if we can fix that nit.

@christianvogt Any ideas on how to fix this approach, I'll try to think something later but now I'm very focused on other tasks, any suggestion is really appreciated.


https://github.com/opendatahub-io/odh-dashboard/assets/16117276/e6650488-efd3-4e32-9398-cb9a3c95e150

**Review by lucferbux — CHANGES_REQUESTED** [review]

File: frontend/src/pages/modelServing/screens/projects/InferenceServiceModal/DataConnectionFolderPathField.tsx (hunk 1 above)
Comment (dpanshug) [reply frontend/src/pages/modelServing/screens/projects/InferenceServiceModal/DataConnectionFolderPathField.tsx:37]: Resolved this by adding cleanup in the useEffect

```
  React.useEffect(() => {
    debouncedValidateFolderPath();

    // Cleanup the previous debounced calls
    return debouncedValidateFolderPath.cancel;
  }, [debouncedValidateFolderPath]);
```

**Comment by dpanshug** [comment]
@vconzola made your suggested changes, pls re-check

File: frontend/src/pages/modelServing/screens/projects/InferenceServiceModal/DataConnectionFolderPathField.tsx (hunk 1 above)
Comment (lucferbux) [reply frontend/src/pages/modelServing/screens/projects/InferenceServiceModal/DataConnectionFolderPathField.tsx:37]: Perfect! that's it!

**Comment by lucferbux** [comment]
Once @vconzola lgtm it we can approve it, great work here!

File: frontend/src/pages/modelServing/screens/projects/InferenceServiceModal/DataConnectionFolderPathField.tsx (hunk 1 above)
Comment (christianvogt) [reply frontend/src/pages/modelServing/screens/projects/InferenceServiceModal/DataConnectionFolderPathField.tsx:37]: @lucferbux @dpanshug the proper way to use `useDebounceCallback` is to ensure that the debounce callback function is a stable reference.
The current implementation uses the fact that the debounce function is unstable and changes every time `folderPath` changes.

```
  const debouncedValidateFolderPath = useDebounceCallback(
    React.useCallback((path: string) => {
      if (path !== '' && (containsOnlySlashes(path) || !isS3PathValid(path))) {
        setValidated('error');
      } else {
        setValidated('default');
      }
    }, []),
  );

  React.useEffect(() => {
    debouncedValidateFolderPath(folderPath);
  }, [debouncedValidateFolderPath, folderPath]);
```

Comment (lucferbux) [reply frontend/src/pages/modelServing/screens/projects/InferenceServiceModal/DataConnectionFolderPathField.tsx:37]: mmmm ok, thanks for the insight!

**Review by lucferbux — APPROVED** [review]
/lgtm

Code looks great, good job here!

**Pull request merged**